tests/test_askrene.py

```diff
@@ -7,23 +7,128 @@
 import pytest
 import time
 
+def direction(src, dst):
+    """BOLT 7 direction: 0 means from lesser encoded id"""
+    if src < dst:
+        return 0
+    return 1
+
+
+def test_reserve(node_factory):
+    """Test reserving channels"""
+    l1, l2, l3 = node_factory.line_graph(3, wait_for_announce=True)
+
+    assert l1.rpc.askrene_listreservations() == {'reservations': []}
+    scid12dir = f"{first_scid(l1, l2)}/{direction(l1.info['id'], l2.info['id'])}"
+    scid23dir = f"{first_scid(l2, l3)}/{direction(l2.info['id'], l3.info['id'])}"
+
+    initial_prob = l1.rpc.getroutes(source=l1.info['id'],
+                                    destination=l3.info['id'],
+                                    amount_msat=1000000,
+                                    layers=[],
+                                    maxfee_msat=100000,
+                                    final_cltv=0)['probability_ppm']
+
+    # Reserve 1000 sats on path.  This should reduce probability!
+    l1.rpc.askrene_reserve(path=[{'short_channel_id_dir': scid12dir,
+                                  'amount_msat': 1000_000},
+                                 {'short_channel_id_dir': scid23dir,
+                                  'amount_msat': 1000_000}])
+    listres = l1.rpc.askrene_listreservations()['reservations']
+    if listres[0]['short_channel_id_dir'] == scid12dir:
+        assert listres[1]['short_channel_id_dir'] == scid23dir
+    else:
+        assert listres[0]['short_channel_id_dir'] == scid23dir
+        assert listres[1]['short_channel_id_dir'] == scid12dir
+    assert len(listres) == 2
+
+    assert l1.rpc.getroutes(source=l1.info['id'],
+                            destination=l3.info['id'],
+                            amount_msat=1000000,
+                            layers=[],
+                            maxfee_msat=100000,
+                            final_cltv=0)['probability_ppm'] < initial_prob
+
+    # Now reserve so much there's nothing left.
+    l1.rpc.askrene_reserve(path=[{'short_channel_id_dir': scid12dir,
+                                  'amount_msat': 1000_000_000_000},
+                                 {'short_channel_id_dir': scid23dir,
+                                  'amount_msat': 1000_000_000_000}])
+
+    # FIXME: better error!
+    with pytest.raises(RpcError, match="Could not find route"):
+        l1.rpc.getroutes(source=l1.info['id'],
+                         destination=l3.info['id'],
+                         amount_msat=1000000,
+                         layers=[],
+                         maxfee_msat=100000,
+                         final_cltv=0)['probability_ppm']
+
+    # Can't remove wrong amounts: that's user error
+    with pytest.raises(RpcError, match="Unknown reservation"):
```

Comment:
Yes, but if the user had had the error on the second path element then the first item in the path
does get unreserved and still I get an RPC error.
Then a subsequent correct call to unreserve will fail because it will not find the first path element.

I don't want to be too pedantic. I think we can leave as it is and in the future make a reserve/unreserve RPC
that works all or nothing.

Reply:
Yes, it's user-error.  We shouldn't try to hard to save them...